File: UISamples/Bit6UIFullDemo-ObjC/ChatDemo/ConversationDetailsTableViewController.m
//
//  ConversationDetailsTableViewController.m
//  Bit6
//
//  Created by Carlos Thurber on 03/11/15.
//  Copyright (c) 2015 Bit6. All rights reserved.
//

#import "ConversationDetailsTableViewController.h"

@interface ConversationDetailsTableViewController ()

@property (nonatomic, strong) Bit6Group *group;
@property (nonatomic, weak) UITextField *subjectTextField;

@end

@implementation ConversationDetailsTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.rightBarButtonItem = self.group.isAdmin?self.editButtonItem:nil;
    self.tableView.tableHeaderView = [[UIView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, self.tableView.bounds.size.width, 0.01f)];
    self.tableView.allowsSelectionDuringEditing = YES;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.navigationController setToolbarHidden:YES animated:YES];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [self.subjectTextField resignFirstResponder];
}

- (void)setConversation:(Bit6Conversation *)conversation
{
    _conversation = conversation;
    self.group = [Bit6Group groupWithConversation:conversation];
}

- (void)setEditing:(BOOL)editing animated:(BOOL)animated
{
    [super setEditing:editing animated:animated];
    
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:self.group.members.count+1 inSection:0];
    if (editing) {
        [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
    }
    else {
        [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        
        [self.group setMetadata:@{@"title":self.subjectTextField.text} completion:^(NSError *error) {
            if (error != nil) {
                NSLog(@"Failed to change the title");
            }
        }];
    }
    
    self.subjectTextField.enabled = self.editing;
}

- (Bit6GroupMember*)memberForIndexPath:(NSIndexPath*)indexPath
{
    if (self.group) {
        Bit6GroupMember *member = self.group.members[indexPath.row-1];
        return member;
    }
    return nil;
}

#pragma mark - TableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.group) {
        if (self.group.hasLeft) {
            return 0;
        }
        else {
            return 1 + //title cell
            self.group.members.count + //members cells
            (self.editing?1:0); //cell to add member
        }
    }
    else {
        return 1;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = nil;
    
    //title
    if (self.group && indexPath.row == 0) {
        cell = [tableView dequeueReusableCellWithIdentifier:@"group_subject"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"group_subject"];
            UILabel *subjectLabel = [[UILabel alloc] initWithFrame:CGRectZero];
            subjectLabel.text = @"Group subject: ";
            [subjectLabel sizeToFit];
            CGRect frame = subjectLabel.frame;
            frame.origin.x = 15;
            frame.origin.y = (44.0f - frame.size.height)/2;
            subjectLabel.frame = frame;
            [cell.contentView addSubview:subjectLabel];
            
            UITextField *subjectTextField = [[UITextField alloc] initWithFrame:CGRectZero];
            subjectTextField.clearButtonMode = UITextFieldViewModeWhileEditing;
            frame = subjectLabel.frame;
            frame.origin.x = CGRectGetMaxX(frame) + 5;
            frame.size.width = self.tableView.frame.size.width - frame.origin.x - 10;
            subjectTextField.frame = frame;
            subjectTextField.placeholder = @"<No Subject>";
            subjectTextField.enabled = self.editing;
            if ([self.group.metadata[@"title"] isKindOfClass:[NSString class]]) {
                NSString *title = self.group.metadata[@"title"];
                if (title.length>0) {
                    subjectTextField.text = title;
                }
            }
            [subjectTextField addTarget:self action:@selector(subjectTextFieldDidChange:) forControlEvents:UIControlEventEditingDidEndOnExit];
            [cell.contentView addSubview:subjectTextField];
            
            self.subjectTextField = subjectTextField;
        }
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    //members
    else if (!self.group || (indexPath.row-1 < (int)self.group.members.count)) {
        cell = [tableView dequeueReusableCellWithIdentifier:@"group_member"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"group_member"];
            cell.detailTextLabel.textColor = [UIColor redColor];
            cell.detailTextLabel.font = [UIFont systemFontOfSize:14.0f];
        }
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    //add member
    else {
        cell = [tableView dequeueReusableCellWithIdentifier:@"add_member"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"add_member"];
            cell.textLabel.textColor = [UIColor colorWithRed:0 green:0.478431 blue:1.0 alpha:1.0];
        }
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (self.group) {
        //title
        if (indexPath.row == 0) {
            
        }
        //members
        else if (indexPath.row-1 < (int)self.group.members.count) {
            Bit6GroupMember *member = [self memberForIndexPath:indexPath];
            cell.textLabel.text = member.address.uri;
            cell.detailTextLabel.text = [member.role isEqualToString:@"admin"]?@"Group Admin":@"";
        }
        //add member
        else {
            cell.textLabel.text = @"+ Add contact";
        }
    }
    else {
        cell.textLabel.text = self.conversation.address.uri;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 44.0f;
}

- (NSIndexPath*)tableView:(UITableView *)tableView willSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row == self.group.members.count+1){
        return indexPath;
    }
    return nil;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self inviteParticipant];
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row == 0) return NO; //title row
    else if ((indexPath.row-1) < self.group.members.count){
        Bit6GroupMember *member = [self memberForIndexPath:indexPath];
        return ![member.address isEqual:Bit6.session.activeIdentity]; //cannot delete himself
    }
    else {
        return NO;
    }
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)aTableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
    // Detemine if it's in editing mode (used to disable swipe-to-delete gesture)
    if (self.tableView.editing) {
        return UITableViewCellEditingStyleDelete;
    }
    return UITableViewCellEditingStyleNone;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    Bit6GroupMember *member = [self memberForIndexPath:indexPath];
    [self.group kickGroupMember:member completion:^(NSArray *members, NSError *error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (!error) {
                [self.tableView reloadData];
            }
        });
    }];
}

- (void)subjectTextFieldDidChange:(UITextField*)textField { }

#pragma mark - Invite

- (void)inviteParticipant
{
    UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Type the friend username to invite" message:nil preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:nil];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault
                                            handler:^(UIAlertAction * action) {
                                                UITextField *userTextField = alert.textFields[0];
                                                NSString *string = userTextField.text;
                                                
                                                Bit6Address *address = [Bit6Address addressWithUsername:string];
                                                
                                                [self.group inviteGroupMemberWithAddress:address role:Bit6GroupMemberRoleUser completion:^(NSArray *members, NSError *error) {
                                                    if (error) {
                                                        dispatch_async(dispatch_get_main_queue(), ^{
                                                            UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Failed to invite users to the group" message:error.localizedDescription preferredStyle:UIAlertControllerStyleAlert];
                                                            [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
                                                            [self.navigationController presentViewController:alert animated:YES completion:nil];
                                                        });
                                                    }
                                                    else {
                                                        [self.tableView reloadData];
                                                    }
                                                }];
                                            }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    
    [self.navigationController presentViewController:alert animated:YES completion:nil];
}

@end

File: Samples/Bit6ChatDemo-Swift/Bit6ChatDemo-Swift/ConversationDetailsTableViewController.swift
//
//  ConversationDetailsTableViewController.swift
//  Bit6ChatDemo-Swift
//
//  Created by Carlos Thurber on 03/31/15.
//  Copyright (c) 2015 Bit6. All rights reserved.
//

import UIKit

class ConversationDetailsTableViewController: UITableViewController {

    var group : Bit6Group!
    var subjectTextField : UITextField!
    
    var conversation : Bit6Conversation! {
        didSet {
            conversation.currentConversation = true
            self.group = Bit6Group(forConversation:conversation)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "Details"
        
        if self.group != nil && self.group.isAdmin {
            self.navigationItem.rightBarButtonItem = self.editButtonItem()
        }
        
        self.tableView.tableHeaderView = UIView(frame: CGRectMake(0, 0, self.tableView.bounds.size.width, 0.001))
        self.tableView.allowsSelectionDuringEditing = true
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setToolbarHidden(true, animated: true)
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        
        if self.subjectTextField != nil {
            self.subjectTextField.resignFirstResponder()
        }
    }
    
    override func setEditing(editing:Bool, animated:Bool) {
        super.setEditing(editing, animated:animated)
        let indexPath = NSIndexPath(forRow: self.group.members.count+1, inSection: 0)
        if editing {
            self.tableView.insertRowsAtIndexPaths([indexPath], withRowAnimation:.Automatic)
        }
        else {
            self.tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation:.Automatic)
            
            let subject = self.subjectTextField.text == nil ? "" : self.subjectTextField.text!
            
            self.group.setMetadata(["title":subject]){ (error) in
                if error != nil {
                    NSLog("Failed to change the title")
                }
            }
        }
        
        self.subjectTextField.enabled = self.editing
    }

    func memberForIndexPath(indexPath:NSIndexPath) -> Bit6GroupMember! {
        if self.group != nil {
            let member = self.group.members[indexPath.row-1] as! Bit6GroupMember
            return member
        }
        return nil
    }
    
    // MARK: - Table view data source

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if self.group != nil {
            if self.group.hasLeft {
                return 0
            }
            else {
                return 1 + //title cell
                    self.group.members.count + //members cells
                    (self.editing ? 1 : 0) //cell to add member
            }
        }
        else {
            return 1
        }
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        //title
        if self.group != nil && indexPath.row == 0 {
            var cell = tableView.dequeueReusableCellWithIdentifier("group_subject") as UITableViewCell!
            if cell == nil {
                cell = UITableViewCell(style: .Default, reuseIdentifier: "group_subject")
                let subjectLabel = UILabel(frame: CGRectZero)
                subjectLabel.text = "Group subject: "
                subjectLabel.sizeToFit()
                var frame = subjectLabel.frame
                frame.origin.x = 15
                frame.origin.y = (44.0 - frame.size.height)/2
                subjectLabel.frame = frame;
                cell.contentView.addSubview(subjectLabel)
                
                let subjectTextField = UITextField(frame: CGRectZero)
                subjectTextField.clearButtonMode = .WhileEditing
                frame = subjectLabel.frame;
                frame.origin.x = CGRectGetMaxX(frame) + 5;
                frame.size.width = self.tableView.frame.size.width - frame.origin.x - 10;
                subjectTextField.frame = frame;
                subjectTextField.placeholder = Bit6EmptyGroupSubject;
                subjectTextField.enabled = self.editing;
                
                let title = self.group.metadata!["title"] as! NSString!
                if let title = title {
                    subjectTextField.text = title as String
                }
                subjectTextField.addTarget(self, action:"subjectTextFieldDidChange:", forControlEvents:.EditingDidEndOnExit)
                cell.contentView.addSubview(subjectTextField)
                
                self.subjectTextField = subjectTextField;
            }
            cell.selectionStyle = .None
            return cell
        }
            
            //members
        else if self.group == nil  || (indexPath.row-1 < self.group.members.count) {
            var cell = tableView.dequeueReusableCellWithIdentifier("group_member") as UITableViewCell!
            if cell == nil {
                cell = UITableViewCell(style: .Subtitle, reuseIdentifier: "group_member")
                cell.detailTextLabel!.textColor = UIColor.redColor()
                cell.detailTextLabel!.font = UIFont.systemFontOfSize(14)
            }
            cell.selectionStyle = .None
            return cell
        }
            
            //add member
        else {
            var cell = tableView.dequeueReusableCellWithIdentifier("add_member") as UITableViewCell!
            if cell == nil {
                cell = UITableViewCell(style: .Default, reuseIdentifier: "add_member")
                cell.textLabel!.textColor = UIColor(red: 0, green: 0.478431, blue: 1.0, alpha: 1.0)
            }
            return cell
        }
    }
    
    override func tableView(tableView: UITableView, willDisplayCell cell: UITableViewCell, forRowAtIndexPath indexPath: NSIndexPath) {
        if self.group != nil {
            //title
            if indexPath.row == 0 {
                
            }
            //members
            else if (indexPath.row-1 < self.group.members.count) {
                let member = self.memberForIndexPath(indexPath)
                cell.textLabel!.text = member.address.displayName
                cell.detailTextLabel!.text = member.role == "admin" ? "Group Admin" : ""
            }
            //add member
            else {
                cell.textLabel!.text = "+ Add contact";
            }
        }
        else {
            cell.textLabel!.text = self.conversation.address.displayName;
        }
    }
    
    override func tableView(tableView:UITableView, heightForRowAtIndexPath indexPath:NSIndexPath) -> CGFloat {
        return 44.0
    }
    
    override func tableView(tableView:UITableView, willSelectRowAtIndexPath indexPath:NSIndexPath) -> NSIndexPath? {
        if indexPath.row == self.group.members.count+1 {
            return indexPath
        }
        return nil
    }
    
    override func tableView(tableView:UITableView, didSelectRowAtIndexPath indexPath:NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated:true)
        self.inviteParticipant()
    }
    
    override func tableView(tableView:UITableView, canEditRowAtIndexPath indexPath:NSIndexPath) -> Bool {
        if indexPath.row == 0 { return false } //title row
        else if (indexPath.row-1) < self.group.members.count {
            let member = self.memberForIndexPath(indexPath)
            return !member.address.isEqual(Bit6.session().userIdentity) //cannot delete himself
        }
        else {
            return false;
        }
    }
    
    override func tableView(tableView:UITableView, editingStyleForRowAtIndexPath indexPath:NSIndexPath) -> UITableViewCellEditingStyle {
        // Detemine if it's in editing mode (used to disable swipe-to-delete gesture)
        if self.tableView.editing {
            return .Delete
        }
        return .None
    }
    
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath){
        let member = self.memberForIndexPath(indexPath)
        self.group.deleteMember(member){ (members, error) in
            if error == nil {
                dispatch_async(dispatch_get_main_queue()) {
                    self.tableView.reloadData()
                }
            }
        }
    }
    
    func subjectTextFieldDidChange(textField:UITextField) { }

    // MARK: - Invite
    
    func inviteParticipant() {
        let alert = UIAlertController(title:"Type the friend username to invite", message: nil, preferredStyle: .Alert)
        alert.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler:{(action :UIAlertAction) in
            
        }))
        alert.addAction(UIAlertAction(title: "Ok", style: .Default, handler:{(action :UIAlertAction) in
            let usernameTextField = alert.textFields![0]
            if usernameTextField.text?.characters.count > 0 {
                
                let destinations =  [usernameTextField.text]
                
                if destinations.count == 1 {
                    
                    let addresses = NSMutableArray(capacity: destinations.count)
                    
                    for (_, element) in destinations.enumerate() {
                        if let address = Bit6Address(kind: .USERNAME, value: element) {
                            addresses.addObject(address)
                        }
                    }
                    
                    self.group.inviteAddresses(addresses as [AnyObject]) { (members, error) in
                        if error != nil {
                            let alert = UIAlertController(title:"Failed to invite users to the group", message: nil, preferredStyle: .Alert)
                            alert.addAction(UIAlertAction(title: "OK", style: .Default, handler:nil))
                            self.navigationController?.presentViewController(alert, animated: true, completion:nil)
                        }
                        else {
                            self.tableView.reloadData()
                        }
                    }
                }
            }
        }))
        alert.addTextFieldWithConfigurationHandler({(textField: UITextField) in
            textField.placeholder = "Username"
        })
        
        self.navigationController?.presentViewController(alert, animated: true, completion:nil)
    }
}
